### PR TITLE
Ensure instance migration can't detach boundary event from element instance

### LINKED ISSUE #19956: Ensure instance migration can't detach boundary event from element instance
Part of #19640 

>Map a message boundary event that is attached to a different element in the target as it was in the source.

### PR DESCRIPTION
**Pull request by korthout**
## Description

Adds an additional validation to instance migration.

If boundary event would become detached from element instance, the `MIGRATE` command is rejected with [`GRPC_STATUS_FAILED_PRECONDITION`](https://docs.camunda.io/docs/next/apis-tools/zeebe-api/gateway-service/#grpc_status_failed_precondition-3).

I've taken the liberty to clean up some precondition error templates.

The `gateway.proto` documentation is adjusted accordingly. We'll also need to update the docs at https://docs.camunda.io.

> [!NOTE]
> The `gateway.proto` file was not completely synced with https://docs.camunda.io. To align them, I've added:
> - a mapping instruction changes the implementation of a task

## Related issues

closes #19956

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java [hunk 1]
```diff
@@ -494,6 +502,47 @@ private static void requireNoBoundaryEvent(
     }
   }
 
+  public static void requireMappedBoundaryEventsToStayAttachedToSameElement(
```
Comment (berkaycanbc) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:521]: 🔧 This naming explains "what will happen as a result (boundary event will be detached wrongly)". A name that explains what the method does would be a better choice. E.g. `requireMappedBoundaryEventsExistInTarget` which explains what the method does. WDYT?

**Review by berkaycanbc — CHANGES_REQUESTED** [review]
Implementation LGTM @korthout! 🚀 

I requested a naming change which I think, is still important to discuss.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java (hunk 1 above)
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:521]: 🤔 But `requireMappedBoundaryEventsExistInTarget` is not correct though. It's not testing whether or not the mapped boundary event exists in the target. It checks that the mapped boundary event stays attached to the same element after mapping, i.e. that the boundary event is not detached from the element it was initially attached to.

Do you see any other options?

Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:521]: Yeah, I think, the name and the error message you chose is more user friendly. It actually indicates something that cannot be done by user when they try to move the boundary event to a different task. E.g. 
Screenshot 2024-07-04 at 19 54 51

**Review by berkaycanbc — APPROVED** [review]

**Pull request merged**